ledger/participant-integration-api/src/main/scala/platform/indexer/MeteringAggregator.scala

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.indexer
+
+import com.daml.ledger.offset.Offset
+import com.daml.ledger.participant.state.index.v2.MeteringStore
+import com.daml.ledger.participant.state.index.v2.MeteringStore.ParticipantMetering
+import com.daml.ledger.resources.ResourceOwner
+import com.daml.lf.data.Time.Timestamp
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.metrics.Metrics
+import com.daml.platform.indexer.MeteringAggregator.{toOffsetDateTime, toTimestamp}
+import com.daml.platform.store.appendonlydao.{DbDispatcher, SqlExecutor}
+import com.daml.platform.store.backend.ParameterStorageBackend.LedgerMeteringEnd
+import com.daml.platform.store.backend.{MeteringStorageWriteBackend, ParameterStorageBackend}
+
+import java.sql.Connection
+import java.time.temporal.ChronoUnit
+import java.time.{OffsetDateTime, ZoneOffset}
+import java.util.{Timer, TimerTask}
+import scala.concurrent.duration._
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+object MeteringAggregator {
+
+  class Owner(
+      meteringStore: MeteringStorageWriteBackend,
+      parameterStore: ParameterStorageBackend,
+      metrics: Metrics,
+      period: FiniteDuration = 6.minutes,
+  ) {
+
+    private[platform] def apply(
+        dispatcher: DbDispatcher
+    )(implicit loggingContext: LoggingContext): ResourceOwner[Unit] = {
+      val aggregator = new MeteringAggregator(meteringStore, parameterStore, metrics)
+      Await.ready(aggregator.initialize(dispatcher), 60.seconds)
```

Comment:
this would mean that we initiate as we create the resource owner (as opposed to initilize as part of resource acquire process)
you need something like this:
```
      for {
        _ <- ResourceOwner.forFuture(() => aggregator.initialize(dispatcher))
        _ <- ResourceOwner.forTimer(() => new Timer()).map { timer =>
          timer.scheduleAtFixedRate(
            new TimerTask {
              override def run(): Unit = {
                Await.ready(aggregator.run(dispatcher), 60.minutes)
                ()
              }
            },
            period.toMillis,
            period.toMillis,
          )
        }
      } yield ()
```

Reply:
This needs to be fixed on this PR because it risks breaking the resource management in the long run.

Reply:
Ok.

Reply:
That is much nicer.